PNP: early FX numeric suite tool to quantize each layer N times

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#80521 PNP: early FX numeric suite tool to quantize each layer N times**

Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config = get_native_backend_config()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

vkuzo committed:
not for land: early FX numeric suite tool to quantize each layer N times

Summary:

This is far from needing a review, just saving early work.

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: early FX numeric suite tool to quantize each layer N times"


Summary:

This is far from needing a review, just saving early work.

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: early FX numeric suite tool to quantize each layer N times"


Summary:

This is far from needing a review, just saving early work.

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
results_grouped = group_results_by_subgraph(results)
results_comparison = create_results_comparison(
    results_grouped, torch.ao.ns.fx.utils.compute_sqnr, 'sqnr')
print_n_shadows_summary(results_comparison)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "not for land: early FX numeric suite tool to quantize each layer N times"


Summary:

This is far from needing a review, just saving early work.

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
results_grouped = group_results_by_subgraph(results)
results_comparison = create_results_comparison(
    results_grouped, torch.ao.ns.fx.utils.compute_sqnr, 'sqnr')
print_n_shadows_summary(results_comparison)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "not for land: early FX numeric suite tool to quantize each layer N times"


Summary:

This is far from needing a review, just saving early work.

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
results_grouped = group_results_by_subgraph(results)
results_comparison = create_results_comparison(
    results_grouped, torch.ao.ns.fx.utils.compute_sqnr, 'sqnr')
print_n_shadows_summary(results_comparison)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "not for land: early FX numeric suite tool to quantize each layer N times"


Summary:

This is far from needing a review, just saving early work.

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
results_grouped = group_results_by_subgraph(results)
results_comparison = create_results_comparison(
    results_grouped, torch.ao.ns.fx.utils.compute_sqnr, 'sqnr')
print_n_shadows_summary(results_comparison)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "not for land: early FX numeric suite tool to quantize each layer N times"


Summary:

This is far from needing a review, just saving early work.

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
results_grouped = group_results_by_subgraph(results)
results_comparison = create_results_comparison(
    results_grouped, torch.ao.ns.fx.utils.compute_sqnr, 'sqnr')
print_n_shadows_summary(results_comparison)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "not for land: early FX numeric suite tool to quantize each layer N times"


Summary:

This is far from needing a review, just saving early work.

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
results_grouped = group_results_by_subgraph(results)
results_comparison = create_results_comparison(
    results_grouped, torch.ao.ns.fx.utils.compute_sqnr, 'sqnr')
print_n_shadows_summary(results_comparison)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "not for land: early FX numeric suite tool to quantize each layer N times"


Summary:

This is far from needing a review, just saving early work.

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
results_grouped = group_results_by_subgraph(results)
results_comparison = create_results_comparison(
    results_grouped, torch.ao.ns.fx.utils.compute_sqnr, 'sqnr')
print_n_shadows_summary(results_comparison)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: early FX numeric suite tool to quantize each layer N times"


Summary:

This is far from needing a review, just saving early work.

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
results_grouped = group_results_by_subgraph(results)
results_comparison = create_results_comparison(
    results_grouped, torch.ao.ns.fx.utils.compute_sqnr, 'sqnr')
print_n_shadows_summary(results_comparison)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: early FX numeric suite tool to quantize each layer N times"


Summary:

This is far from needing a review, just saving early work.

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
results_grouped = group_results_by_subgraph(results)
results_comparison = create_results_comparison(
    results_grouped, torch.ao.ns.fx.utils.compute_sqnr, 'sqnr')
print_n_shadows_summary(results_comparison)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "not for land: early FX numeric suite tool to quantize each layer N times"


Summary:

This is far from needing a review, just saving early work.

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
results_grouped = group_results_by_subgraph(results)
results_comparison = create_results_comparison(
    results_grouped, torch.ao.ns.fx.utils.compute_sqnr, 'sqnr')
print_n_shadows_summary(results_comparison)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
results_grouped = group_results_by_subgraph(results)
results_comparison = create_results_comparison(results_grouped)
print_n_shadows_summary(results_comparison)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config_dict = get_native_backend_config_dict()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config_dict)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

jerryzh168 commented:
(I updated backend_config_dict to backend_config in the summary)

jerryzh168 commented on test/quantization/fx/test_numeric_suite_fx.py:
```diff
@@ -2060,6 +2070,223 @@ def forward(self, x):
+                x = torch.cat((x,))
+                x = torch.cat(tensors=[x])
+                # TODO(future PR): enable layernorm
+                # blocked on FX graph mode quant not inserting observer for
```
is this fixed now? with https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/backend_config/_common_operator_config_utils.py#L355

jerryzh168 commented on torch/ao/ns/_numeric_suite_fx.py:
```diff
@@ -189,12 +207,24 @@ def __init__(
     def forward(self, x):
         """
         """  # blank docblock to make autodoc happy
+        if not self.enabled:
+            return x
+        if not self.save_activations:
+            return x
```
are these two flags doing the same thing?

vkuzo replied:
technically no, but I can see how the difference is not explained well.  `self.enabled` is controlling if the logger is doing anything at all, and `save_activations` is only controlling whether the logger is saving activations.  A logger could be enabled and be saving comparisons, but not activations. I'll clean this up.

jerryzh168 commented on torch/ao/ns/_numeric_suite_fx.py:
```diff
@@ -210,6 +240,36 @@ def __repr__(self):
+        if not self.enabled:
+            return x
+        assert isinstance(x, torch.Tensor), 'non-tensor inputs not yet supported'
+        if self.save_activations:
+            # save the activation, for debugging
+            self.stats.append(x.detach())
```
why does `OutputComparisonLogger` also save activation? should we use OutputLogger to save activation instead?

jerryzh168 replied:
maybe we can run the forward of `OutputLogger` here?

jerryzh168 commented on torch/ao/ns/_numeric_suite_fx.py:
```diff
@@ -673,3 +741,113 @@ def extend_logger_results_with_comparison(
+    model: torch.nn.Module,
+    example_inputs: Any,
+    qconfig_mappings: List[QConfigMapping],
+    backend_config_dict: Any,
```
nit: backend_config_dict --> backend_config

jerryzh168 commented on torch/ao/ns/_numeric_suite_fx.py:
```diff
@@ -673,3 +741,113 @@ def extend_logger_results_with_comparison(
+
+def convert_n_shadows_model(model: GraphModule) -> GraphModule:
+    for node in model.graph.nodes:
+        if node.name.startswith(SHADOW_WRAPPER_NODE_NAME_PREFIX):
```
`name` is not very reliable, can we check the type of the module?

vkuzo replied:
added a TODO for this

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -0,0 +1,893 @@
+    """
+    Output propagation (modeled from shape propagation).
+
+    Given a GraphModule and an example input, saves the input flowing
```
is this `input` or `output`?

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -0,0 +1,893 @@
+    # Dict items are not reversible until Python 3.8, so we hack it
+    # to be compatible with previous Python versions
+    matches_items_reversed: List[Tuple[str, MatchResult]] = []
+    for name, cur_match in matches.items():
+        matches_items_reversed.insert(0, (name, cur_match))
```
nit: maybe just
```
matches_items_reversed = reversed(list(matches.items()))
```

vkuzo replied:
I agree with the feedback, let's improve in a separate PR, hope that's OK

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -0,0 +1,893 @@
+            #   'relu': (relu, [(bn, conv), relu], ...),
+            # }
+            #
+            # Ideally we should clean up the `find_matches` function to make
```
yeah, was hoping that we can use subgraph matcher/rewriter so we don't need to maintain this code..

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -0,0 +1,893 @@
+            if isinstance(cur_match[1][0], Node) and isinstance(cur_match[1][1], Node):
+                # (a, b)
+                list_of_nodes = cur_match[1]
+            elif isinstance(cur_match[1][0], tuple):
+                # ((a, b), c)
+                node_a, node_b = cur_match[1][0]
+                node_c = cur_match[1][1]
+                list_of_nodes = _order_nodes(node_a, node_b, node_c)
+            elif isinstance(cur_match[1][1], tuple):
+                # (a, (b, c))
+                node_a, node_b = cur_match[1][1]
+                node_c = cur_match[1][0]
+                list_of_nodes = _order_nodes(node_a, node_b, node_c)
```
these two represent different graphs,
((a, b), c) means:
```
c  ---- a
b  --- /
```

and (a, (b, c)) means:
```
c --> b --> a
```

jerryzh168 replied:
can we flatten the first case to be a list of ops here?

jerryzh168 replied:
oh sorry, I'm not sure what `((a, b), c)` means, is this a part of a larger pattern?

vkuzo replied:
I agree that this is confusing, let's solve in a future PR since this is a local discussion and doesn't really affect the overall design.  Planning to land with this unsolved to unblock other people, hope that's ok.

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -0,0 +1,893 @@
+    elif isinstance(prev_node, tuple):
+        example_inputs = (x.traced_result for x in prev_node)  # type: ignore[assignment]
+    else:
+        # currently some customer models do not have a traced_result in
```
does this happen when the output is not a Tensor? I saw we only save Tensor outputs right now in OutputProp

vkuzo replied:
thanks for the tip! Added a TODO to look into this

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -0,0 +1,893 @@
+    found_at_least_one_qconfig = False
+    for subgraph_candidate_idx in range(len(qconfig_mappings)):
+        node_name_to_qconfig = \
+            list_of_node_name_to_qconfig[subgraph_candidate_idx - 1]
```
wondering what does `subgraph_candidate_idx` - 1 mean here?

vkuzo replied:
What @andrewor14  said is how it works, I'll see how to make it clearer in the code.

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -0,0 +1,893 @@
+    Given a subgraph in `mt` and a subgraph candidate idx, inserts the
+    subgraph candidate copy and instruments it with loggers.
+
+    If subgraph_idx is 0, this is the baseline fp32 subgraph and we just
```
is this supposed to be `subgraph_idx` or `subgraph_candidate_idx`?
also maybe we can change the name for `subgraph_candidate_idx`, I feel it's a bit confusing

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -0,0 +1,893 @@
+                            new_args.append(inner_arg)
+
+            new_kwargs = {}
+            for name, old_kwarg in first_node.kwargs.items():
+                if isinstance(old_kwarg, Node):
+                    new_kwargs[name] = old_kwarg
+                elif isinstance(old_kwarg, (list, tuple)) and len(old_kwarg):
+                    for inner_old_kwarg in old_kwarg:
+                        new_args.append(inner_old_kwarg)
```
also not sure what this does, can we just pass around `first_node.kwargs` directly?

vkuzo replied:
happy to address in a future PR

jerryzh168 reviewed (approved):
looks good, I reviewed the main part, just had some clarification question inline and would be good to add some more comments for these things

andrewor14 reviewed (approved):
Looks great! Echoing Jerry, more comments in general would make it easier to follow.

vkuzo commented:
thanks folks! I will request review again after addressing the comments, since there is a non trivial amount of comments.

vkuzo commented on torch/ao/ns/_numeric_suite_fx.py:
```diff
@@ -210,6 +240,36 @@ def __repr__(self):
+            return x
+        assert isinstance(x, torch.Tensor), 'non-tensor inputs not yet supported'
+        if self.save_activations:
+            # save the activation, for debugging
```
This is the only way the original NS logger (not for PNP, for the other APIs) works, so it's debugging only for the PNP use case.  I'll think about what we can do to make it more clear.

vkuzo commented on torch/ao/ns/_numeric_suite_fx.py:
```diff
@@ -673,3 +741,113 @@ def extend_logger_results_with_comparison(
+       |                           |
+       | -> op0_q_0 -> log_0_1     | -> op1_q_0 -> log_1_1
+       |                           |
+       | -> op0_q_1 -> log_0_1     | -> op1_q_1 -> log_1_2
```
you're right, this needs an update, will do

vkuzo commented on torch/ao/ns/_numeric_suite_fx.py:
```diff
@@ -673,3 +741,113 @@ def extend_logger_results_with_comparison(
+    2. figure out a better way to name the output structure
+    3. for now, results are manually printed in an easy to read way. A logical
+       next step would be to automaticaly output the best qconfig, and then
+       automatically output a single model quantized with that best qconfig.
```
sounds good

vkuzo commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -0,0 +1,893 @@
+        fqn,  # fqn
+        qconfig_str,
+    )
+    logger_mod_orig.enabled = False
```
Yes, exactly

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config = get_native_backend_config()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config = get_native_backend_config()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -0,0 +1,893 @@
+
+        for node in self.graph.nodes:
+            if node.op == 'placeholder':
+                result = next(args_iter)
```
This part is unchanged from https://pytorch.org/docs/stable/fx.html#the-interpreter-pattern, TBH I didn't test this in detail in the interest of time.  Let me know if I should follow up on this!

vkuzo commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -0,0 +1,893 @@
+
+def create_results_comparison(
+    results_grouped,
+) -> Any:
```
Yeah, agreed.  In general I'd expect the `create_results_comparison` API to go through significant changes, this is just a temporary hack to enable early testing.  We will need to redesign the entire thing.

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config = get_native_backend_config()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config = get_native_backend_config()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config = get_native_backend_config()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config = get_native_backend_config()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).

vkuzo committed:
Update on "PNP: early FX numeric suite tool to quantize each layer N times"


Summary:

This PR is an early prototype of a tool to quantize each layer of a model
N times, with N qconfigs each. We follow the design agreed upon in
https://fburl.com/gdoc/e1gaq3ih .

Current API:

```
m = M().eval()
example_input = (torch.randn(2, 2),)
qconfig_mappings = [
    QConfigMapping().set_global(torch.quantization.default_qconfig),
    QConfigMapping().set_global(torch.quantization.default_dynamic_qconfig),
]
backend_config = get_native_backend_config()

msp = prepare_n_shadows_model(
    m, example_input, qconfig_mappings, backend_config)

for _ in range(2):
    msp(*example_input)

msq = convert_n_shadows_model(msp)
msq(*example_input)

results = extract_results_n_shadows_model(msq)
print_comparisons_n_shadows_model(results)

// example output

subgraph_idx    ref_node_name      best_idx        1        2
--------------  ---------------  ----------  -------  -------
subgraph_0      fc1                       2  42.0834  42.6279
subgraph_1      fc2                       2  43.7259  50.0593
```

Test plan:

```
python test/test_quantization.py -k test_n_shadows
```

Differential Revision: [D37650332](https://our.internmc.facebook.com/intern/diff/D37650332)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37650332).
